### ISSUE TITLE
HALOGEN_PROMPT_CACHE=2 deterministically returns stale unrelated answer for a short chat request

### ISSUE THREAD

**Issue by lev-medien-sandkasten** [peonist-ai/halogen-flash-server#65]
## Summary

On the published `0.11.1` image, the mode-2 prompt-cache serving configuration returns the same unrelated answer for a short, deterministic chat-completions request. The same exact request is correct under the documented exact-control configuration (`HALOGEN_PROMPT_CACHE=1`, composable context off).

This is a raw HTTP reproducer: no agent harness, tools, conversation history, retries, or client SDK are involved.

## Environment

- Image: `ghcr.io/peonist-ai/halogen-flash-server:0.11.1`
- API and engine both report `0.11.1` in `/health`
- Model: `halogen-qwen3.8-flash-next`
- One sequential client, no concurrent requests
- `temperature: 1.0`, `seed: 0`, `max_tokens: 32768`

## Minimal request

Send this body to `/v1/chat/completions`:

```json
{
  "model": "halogen-qwen3.8-flash-next",
  "messages": [
    {
      "role": "system",
      "content": "Solve from the supplied evidence only. If the problem is underdetermined, answer INSUFFICIENT. Keep your working internal; do not print derivations, scratchpad, or repeated checks. Respond tersely, with at most one brief line before the final answer. If you include that brief line, put the final answer on a separate last line. The last line must be exactly: FINAL: <answer>.\""
    },
    {
      "role": "user",
      "content": "A deterministic process obeys L=X AND (NOT Y), M=L OR Y, and N=M AND X. Actual inputs are X=1 and Y=0. Under do(Y=1), keep X fixed and recompute. Return L, M, and N as L=value,M=value,N=value."
    }
  ],
  "temperature": 1.0,
  "max_tokens": 32768,
  "seed": 0,
  "stream": false
}
```

Expected final content:

```text
FINAL: L=0,M=1,N=1
```

## Results

| Arm | Server configuration | Result |
|---|---|---:|
| A | `HALOGEN_PROMPT_CACHE=2`, `HALOGEN_COMPOSABLE_CONTEXT=1` | 0/10 correct |
| B | `HALOGEN_PROMPT_CACHE=1`, `HALOGEN_COMPOSABLE_CONTEXT=0` | 10/10 correct |
| A′ | restored A configuration after a full server restart | 0/10 correct |

Every request in all arms returned HTTP 200 and `finish_reason: "stop"`. There were no transport failures or empty finals.

A and A′ always returned this unrelated final:

```text
FINAL: No prior thinking exists to reproduce; the conversation's first turn contains no solution or reasoning trace.
```

The initial request in both A and A′ was made immediately after a full server restart and reported no `cached_tokens`; subsequent mode-2 requests reported 156 cached prompt tokens. The exact-control requests reported no cached prompt tokens.

A direct `curl --data-binary @minimal-request.json` confirmation against the restored mode-2 configuration reproduces the same unrelated answer.

## Scope and current interpretation

The full request is only 161 prompt tokens, below our configured composable-context floor of 2,048 tokens. Actual composable-context chunk reuse should therefore be ineligible for this request. This points to the mode-2 serving/cache path, but I do not know the internal mechanism and do not want to overstate it as a confirmed stale-KV diagnosis.

Is this expected behavior for mode 2? If not, the request above should be a compact regression test.

**Comment by wenis** [comment]
Thanks for the careful, self-contained reproducer. That is exactly the right shape to work from.

I ran your exact request against the live published `0.11.1` image (id `6e07cc756a3a`, the same digest `:latest` points at) and could not reproduce the stale answer under any of the configurations, including your arm A:

- `PROMPT_CACHE=2` + `COMPOSABLE_CONTEXT=1`, `PROMPT_CACHE=2` alone, `PROMPT_CACHE=1` + `COMPOSABLE_CONTEXT=1`, and the exact control all return `FINAL: L=0,M=1,N=1`, cold and warm, byte-identical request (trailing quote and all), 5+ reps each.
- A cross-request battery under mode 2 (a "reproduce prior thinking" planter, an unrelated persona, and a same-system/different-user turn, then the puzzle): mode 2 isolated every conversation correctly, and the puzzle was never contaminated.

Two things in your report point at a configuration difference rather than a defect I can see, and would let me pin it down.

1. Your `cached prompt tokens` is 156. On my server the same request renders to 194 tokens with thinking on, and about 155 with thinking off. So your server is rendering this request in the thinking-off shape, even though the request sets no `reasoning_effort`. That means a server-side thinking default is in play that is not in the report. Could you share the full `/health` output (the whole `server_defaults` block, `cache_mode`, and the thinking/effort/answer-room fields)?

2. Do you set `HALOGEN_CACHE_DIR`? If the prompt cache is persisted to disk (0.10.0+), a snapshot from an earlier session could survive your "full server restart" and be restored for this request. That would explain a stale answer that talks about a prior conversation, and why a restart did not clear it. If so, clearing that directory (or unsetting the flag) would confirm it.

Also helpful: the complete raw response JSON for one failing request (the `usage`/`timings` and any `reasoning` content), your full container env / run command, and your image digest (`podman image inspect --format '{{.Id}}' ...`). The stale text you are seeing ("no prior thinking to reproduce", "the conversation's first turn") is genuinely the model answering a different prompt, one about reproducing prior reasoning, so something is feeding it a context that is not this request. The above should tell us what.

In the meantime, `PROMPT_CACHE=1` (which you found works) is the exact-answer path and a safe fallback.

**Comment by lev-medien-sandkasten** [comment]
Thanks — I repeated the *exact* wire request against our live service, and it still returns the unrelated answer under mode 2:

```text
HTTP 200; finish_reason: stop
content: FINAL: No prior thinking exists to reproduce; the conversation's first turn contains no solution or reasoning trace.
```

The raw response fields requested (with no credentials or unrelated headers) were:

```json
{
  "usage": {
    "prompt_tokens": 161,
    "completion_tokens": 162,
    "total_tokens": 323,
    "prompt_tokens_details": {"cached_tokens": 156},
    "completion_tokens_details": {"reasoning_tokens": 139}
  },
  "timings": {
    "prompt_n": 5,
    "predicted_n": 162,
    "prompt_ms": 55.5,
    "predicted_ms": 4076.9,
    "cache_n": 156,
    "draft_n": 111,
    "draft_n_accepted": 50
  }
}
```

Current `/health` confirms API/engine `0.11.1`, mode-2 prompt cache, composable context enabled with floor 2048, and these relevant fields:

```json
{
  "prompt_cache": {"enabled": true, "cap_mb": 110, "mode": 2, "snapshot_align": 0},
  "composable_context": {"enabled": true, "unit_floor_tokens": 2048, "store_bytes": 4294967296},
  "reasoning_effort_default": "medium",
  "thinking_answer_room": "max(1024, 15% of max_tokens)",
  "server_defaults": {"HALOGEN_REASONING_EFFORT": "medium"}
}
```

The running image ID is `6e07cc756a3a07d85490dafb3714d8e942e6cbb817e4b046022f84383e733464`, matching the ID you reported. Its relevant runtime environment is:

```text
HALOGEN_PROMPT_CACHE=2
HALOGEN_COMPOSABLE_CONTEXT=1
HALOGEN_COMPOSABLE_CONTEXT_FLOOR=2048
HALOGEN_COMPOSABLE_CONTEXT_BYTES=4294967296
HALOGEN_REASONING_EFFORT=medium
```

`HALOGEN_CACHE_DIR` is **not** set in either the unit or container environment.

**Update — clean restart test:** I then inspected `/cache` before restarting. It reported disk persistence disabled (`"disk":{"on":false,"records":0,...}`), so there was no configured or active persisted cache to clear. I restarted the service, waited for `/health`, and sent this exact chat request as the **first** post-restart completion request. It still returned the same unrelated final above. This time the response had no `cached_tokens`, and its timings reported `prompt_n: 161` and `cache_n: 0` (with `prompt_ms: 1664.6`).

That rules out both a surviving in-memory entry and an active on-disk prompt-cache snapshot as the explanation for this reproduction. It does not by itself isolate the remaining mechanism: a mode-2/configuration interaction is supported by the observation, but remains a hypothesis rather than a demonstrated stale-KV cause.

**Comment by wenis** [comment]
Reproduced, thank you. With `reasoning_effort: medium` (your server default, which I had not been setting) I get your exact stale answer under mode 2. Here is what it is and is not.

**Isolation** (published `0.11.1`, your byte-identical request, seed 0):

| config | temp 0 (greedy) | temp 1.0 |
|---|---|---|
| `PROMPT_CACHE=2` (+ CC on or off) | `FINAL: L=0,M=1,N=1` | the stale answer |
| `PROMPT_CACHE=1` | `FINAL: L=0,M=1,N=1` | `FINAL: L=0,M=1,N=1` |

Two things fall out:

1. **It is `PROMPT_CACHE=2`, not composable context.** `PROMPT_CACHE=2` with `COMPOSABLE_CONTEXT=0` still reproduces it; `PROMPT_CACHE=1` with `COMPOSABLE_CONTEXT=1` does not. Composable context is not involved (your request is well under the 2048 floor, as you noted).

2. **It is not stale KV, and not wrong cached content.** It reproduces cold (`cache_n=0`, first request after restart), which you already saw, and at `temperature 0` mode 2 returns the correct answer. If the cache were serving a wrong context, greedy would be wrong too.

The mechanism is the documented one: mode 2 (resume-anywhere) is not bitwise-identical to mode 1. It trades that exactness for the ability to resume from any point, so its logits differ from mode 1 by a tiny amount. At `temperature 1.0` the sampler draws from that slightly-shifted distribution with your seed and lands on a different token early, which for this particular prompt derails into a coherent but wrong answer. Greedy takes the top token in both modes, so it agrees.

So it is expected under mode 2 at high temperature rather than a stale-KV defect, and your own workaround is the right one: `PROMPT_CACHE=1` gives bitwise-exact answers, and `temperature 0` also removes the divergence. One thing we do want to look at: the cold first request has nothing to resume, so in principle its mode-2 path could be made identical to mode 1, which would stop that first-request case from diverging at all. That is an engine change on our side, tracked from here.

**Comment by wenis** [comment]
Shipped in 0.11.3.

What it was: under `HALOGEN_PROMPT_CACHE=2` the server saves its place at the end of the system prompt and at the end of the history, and to save it there it used to split the prompt's forward pass at that point. A split pass is not the same arithmetic as a single one, so a request with nothing to resume still got slightly different logits from the mode-1 answer, and at `temperature 1.0` your prompt turned that into the wrong answer every time. The pass now runs unsplit and the state is captured mid-pass instead.

Your exact request, on the release image, your configuration (`PROMPT_CACHE=2`, composable context on, `reasoning_effort medium`, `temperature 1.0`, `seed 0`): the cold request returns `FINAL: L=0,M=1,N=1`, byte-identical in content to mode 1's, and repeats of it return the same bytes again (the server now keeps the state at the end of the last request, so an exact repeat reads nothing again). The gate behind it checks 161 to 32,768-token prompts for bitwise-identical logits against the cache-off run.

What did not change: a request that resumes from the cache and continues past it (a real next turn) is still a numeric seam, as documented under *Choosing a cache mode*; it is measured no larger than before. `PROMPT_CACHE=1` remains the exact setting for everything. Two visible details: `cached_tokens` on a repeat now reads the whole prompt (161), and `/health` reports `snapshot_align: 64`.

Thank you for the reproducer; it was the whole diagnosis.
